src/choleskyQR.jl

```diff
@@ -51,143 +51,200 @@ function cholesky_jacobimatrix(W::AbstractMatrix, Q)
-    return CholeskyJacobiBand{:ev,T}(ev, U, UX, 2)
+function CholeskyJacobiData(U::AbstractMatrix{T}, UX) where T
+    dv = Vector{T}(undef,2) # compute a length 2 vector on first go
+    ev = Vector{T}(undef,2)
+    dv[1] = UX[1,1]/U[1,1] # this is dot(view(UX,1,1), U[1,1] \ [one(T)])
+    dv[2] = -U[1,2]*UX[2,1]/(U[1,1]*U[2,2])+UX[2,2]/U[2,2] # this is dot(view(UX,2,1:2), U[1:2,1:2] \ [zero(T); one(T)])
+    ev[1] = -UX[1,1]*U[1,2]/(U[1,1]*U[2,2])+UX[1,2]/U[2,2] # this is dot(view(UX,1,1:2), U[1:2,1:2] \ [zero(T); one(T)])
+    ev[2] = UX[2,1]/U[3,3]*(-U[1,3]/U[1,1]+U[1,2]*U[2,3]/(U[1,1]*U[2,2]))+UX[2,2]/U[3,3]*(-U[2,3]/U[2,2])+UX[2,3]/U[3,3] # this is dot(view(UX,2,1:3), U[1:3,1:3] \ [zeros(T,2); one(T)])[1:3,1:3] \ [zeros(T,2); one(T)])
+    return CholeskyJacobiData{T}(dv, ev, U, UX, 2)
 end
 
-size(::CholeskyJacobiBand) = (ℵ₀,) # Stored as an infinite cached vector
+size(::CholeskyJacobiData) = (ℵ₀,2) # Stored as two infinite cached bands
+
+function getindex(K::CholeskyJacobiData, n::Integer, m::Integer)
+    @assert (m==1) || (m==2)
+    resizedata!(K,n,m)
+    m == 1 && return K.dv[n]
+    m == 2 && return K.ev[n]
+end
 
 # Resize and filling functions for cached implementation
-function resizedata!(K::CholeskyJacobiBand, nm::Integer)
+function resizedata!(K::CholeskyJacobiData, n::Integer, m::Integer)
+    nm = max(n,m)
     νμ = K.datasize
     if nm > νμ
-        resize!(K.data,nm)
-        cache_filldata!(K, νμ:nm)
+        resize!(K.dv,nm)
+        resize!(K.ev,nm)
+        _fillcholeskybanddata!(K, νμ:nm)
         K.datasize = nm
     end
     K
 end
-function cache_filldata!(J::CholeskyJacobiBand{:dv,T}, inds::UnitRange{Int}) where T
-    # pre-fill U and UX to prevent expensive step-by-step filling in of cached U and UX in the loop
-    getindex(J.U,inds[end]+1,inds[end]+1)
-    getindex(J.UX,inds[end]+1,inds[end]+1)
 
-    ek = [zero(T); one(T)]
-    @inbounds for k in inds
-        J.data[k] = dot(view(J.UX,k,k-1:k), J.U[k-1:k,k-1:k] \ ek)
-    end
-end
-function cache_filldata!(J::CholeskyJacobiBand{:ev, T}, inds::UnitRange{Int}) where T
+function _fillcholeskybanddata!(J::CholeskyJacobiData{T}, inds::UnitRange{Int}) where T
     # pre-fill U and UX to prevent expensive step-by-step filling in of cached U and UX in the loop
-    getindex(J.U,inds[end]+1,inds[end]+1)
-    getindex(J.UX,inds[end]+1,inds[end]+1)
+    resizedata!(J.U,inds[end]+1,inds[end]+1)
+    resizedata!(J.UX,inds[end]+1,inds[end]+1)
 
-    ek = [zeros(T,2); one(T)]
+    dv, ev, UX, U = J.dv, J.ev, J.UX, J.U
     @inbounds for k in inds
-        J.data[k] = dot(view(J.UX,k,k-1:k+1), J.U[k-1:k+1,k-1:k+1] \ ek)
+        # this is dot(view(UX,k,k-1:k), U[k-1:k,k-1:k] \ ek)
+        dv[k] = -U[k-1,k]*UX[k,k-1]/(U[k-1,k-1]*U[k,k])+UX[k,k]./U[k,k]
+        ev[k] = UX[k,k-1]/U[k+1,k+1]*(-U[k-1,k+1]/U[k-1,k-1]+U[k-1,k]*U[k,k+1]/(U[k-1,k-1]*U[k,k]))+UX[k,k]/U[k+1,k+1]*(-U[k,k+1]/U[k,k])+UX[k,k+1]/U[k+1,k+1]  
     end
+    J.dv[inds] .= dv[inds]
```

Comment:
Why is this needed? `dv == J.dv` so this is copying the same data to istelf. Same with the next line.